### PR TITLE
Allow user to specify substitution function in pcustom struct

### PR DESCRIPTION
**Pull request by daniellimws**
### Description/Motivation/Screenshots ###

Currently the user can specify how to substitute values in a struct using `_values_` in the pcustom struct file.

In addition to the current ability to substitute values using a list of tuples, after this patch the user can also specify a function to do the substitution,  giving more flexibility.

An example of this being used is in the [socketaddr_in](https://github.com/hugsy/gef-extras/pull/8/files#diff-949c9b64b64de1131b192b7bd065b4ae) struct.

https://asciinema.org/a/0iygyS51V8j6dLmft243wABGq

### How Has This Been Tested? ###
Tested using the [socketaddr_in](https://github.com/hugsy/gef-extras/pull/8/files#diff-949c9b64b64de1131b192b7bd065b4ae) struct.

| Architecture | Yes/No                   | Comments               |
|--------------|:------------------------:|------------------------|
| x86-32       | :heavy_multiplication_x: | Replace with :heavy_check_mark: if tested |
| x86-64       | :heavy_check_mark: |                        |
| ARM          | :heavy_multiplication_x: |                        |
| AARCH64      | :heavy_multiplication_x: |                        |
| MIPS         | :heavy_multiplication_x: |                        |
| POWERPC      | :heavy_multiplication_x: |                        |
| SPARC        | :heavy_multiplication_x: |                        |

### Checklist ###

- [x] My code follows the code style of this project.
- [ ] My change includes a change to the documentation, if required.
- [x] I have read and agree to the **CONTRIBUTING** document.
- [ ] I have tagged the PR as appropriate (e.g. bug fix).

### CODE REVIEW

**Comment by daniellimws** [comment]
Would you like me to update the docs on this as well?

**Comment by Grazfather** [comment]
Brilliant.

I am not sure if testing `type(...) == list` is the best way to determine if it's a list, though. Perhaps we should invert the logic and instead do something like

```python
if callable(...):
    ...
else:
    try:
        # assume it's a list
    except:
        # error here
```

File: gef.py [hunk 1]
```diff
@@ -4231,9 +4231,12 @@ def get_ctypes_value(self, struct, item, value):
         default = ""
         for name, values in values_list:
             if name != item: continue
-            for val, desc in values:
-                if value == val: return desc
-                if val is None: default = desc
+            if type(values) == list:
```
Comment (Grazfather) [inline gef.py]: You're testing that it's a list, but don't test that it's a list of 2-tuples. I think that gives extra reason to instead just test that it's callable and catch if it later on doesn't fit `[(a,b),...]`

**Comment by daniellimws** [comment]
Alright fixed

**Comment by Grazfather** [comment]
Tested. Looks great.

```c
#include <stdio.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

int main(void) {

	struct sockaddr_in myaddr;
	int s;

	myaddr.sin_family = AF_INET;
	myaddr.sin_port = htons(8888);
	inet_aton("127.0.0.1", (struct in_addr*)&myaddr.sin_addr.s_addr);

	printf("%p\n", &myaddr);

	return 0;
}
```

**Pull request merged**